crates/flux-infer/src/infer.rs

```diff
@@ -494,14 +510,11 @@ impl Sub {
                 debug_assert_eq!(uint_ty_a, uint_ty_b);
                 Ok(())
             }
-            (BaseTy::Adt(adt_a, args_a), BaseTy::Adt(adt_b, args_b)) => {
-                debug_assert_eq!(adt_a.did(), adt_b.did());
-                debug_assert_eq!(args_a.len(), args_b.len());
-                let variances = infcx.genv.variances_of(adt_a.did());
-                for (variance, ty_a, ty_b) in izip!(variances, args_a.iter(), args_b.iter()) {
-                    self.generic_args(infcx, *variance, ty_a, ty_b)?;
-                }
-                Ok(())
+            (BaseTy::Adt(a_adt, a_args), BaseTy::Adt(b_adt, b_args)) => {
+                self.btys_def_id_args(infcx, a_adt.did(), a_args, b_adt.did(), b_args)
+            }
+            (BaseTy::FnDef(a_def_id, a_args), BaseTy::FnDef(b_def_id, b_args)) => {
+                self.btys_def_id_args(infcx, *a_def_id, a_args, *b_def_id, b_args)
```

Comment:
This is related to how we handle generic arguments. Since they are always going to be unrefined maybe better to have a stricter assertion until we have a concrete example.

```suggestion
                debug_assert_eq!(a_def_id, b_def_id);
                assert_eq!(a_args, b_args);
                Ok(())
```

The first one is `debug_assert` because I'm confident that it will never fail. The second one we want to catch when so we have a concrete example to fix.